game-core/src/test/java/games/strategy/triplea/delegate/battle/steps/BattleStepsTest.java

```diff
@@ -0,0 +1,2199 @@
+import static games.strategy.triplea.delegate.battle.BattleStepStrings.SELECT_PREFIX;
+import static games.strategy.triplea.delegate.battle.BattleStepStrings.SUBMERGE_SUBS_VS_AIR_ONLY;
+import static games.strategy.triplea.delegate.battle.BattleStepStrings.SUBS_SUBMERGE;
+import static games.strategy.triplea.delegate.battle.BattleStepStrings.SUBS_WITHDRAW;
+import static games.strategy.triplea.delegate.battle.steps.BattleSteps.FIRE;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import games.strategy.engine.data.GameData;
+import games.strategy.engine.data.GamePlayer;
+import games.strategy.engine.data.Territory;
+import games.strategy.engine.data.Unit;
+import games.strategy.engine.data.UnitType;
+import games.strategy.engine.data.properties.GameProperties;
+import games.strategy.triplea.Constants;
+import games.strategy.triplea.attachments.TechAttachment;
+import games.strategy.triplea.attachments.UnitAttachment;
+import java.util.Collection;
+import java.util.List;
+import java.util.Set;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+class BattleStepsTest {
+
+  @Mock GameData mockGameData;
+  @Mock GameProperties mockGameProperties;
+  @Mock Function<Collection<Unit>, Collection<Unit>> mockGetDependentUnits;
+  @Mock Supplier<Collection<Territory>> getAttackerRetreatTerritories;
+
+  @Mock
+  BiFunction<GamePlayer, Collection<Unit>, Collection<Territory>> getEmptyOrFriendlySeaNeighbors;
+
+  @Mock Territory mockBattleSite;
+  @Mock GamePlayer attacker;
+  @Mock GamePlayer defender;
+  @Mock TechAttachment techAttachment;
+  @Mock Unit unit1;
+  @Mock UnitType mockUnit1Type;
+  @Mock UnitAttachment mockUnit1Attachment;
+  @Mock Unit unit2;
+  @Mock UnitType mockUnit2Type;
+  @Mock UnitAttachment mockUnit2Attachment;
+  @Mock Unit unit3;
+  @Mock UnitType mockUnit3Type;
+  @Mock UnitAttachment mockUnit3Attachment;
+  @Mock Unit unit4;
+  @Mock UnitType mockUnit4Type;
+  @Mock UnitAttachment mockUnit4Attachment;
+
+  @BeforeEach
+  void setupMocks() {
+    when(mockGameData.getProperties()).thenReturn(mockGameProperties);
+  }
+
+  private void players() {
+    when(attacker.getName()).thenReturn("mockAttacker");
+    when(defender.getName()).thenReturn("mockDefender");
+  }
+
+  private void unit1() {
```

Comment:
Explicit input/output of test setup methods can help make for more explicit unit tests. This method for example could possibly be changed to:

```
private Unit givenUnitWithAttachment(UnitAttachment unitAttachment) {
  Unit unit = mock(Unit.class);
  UnitType unitType = mock(UnitType.class);
  when(unit.getType).thenReturn(unitType);
  when(unitType.getAttachment(anyString()).thenReturn(unitAttachment);
  return unit;
}
```

Then a test, for at least this, can be a bit more self-contained, eg:

```
    players();
    attackerRetreat();
    Unit unit1 = givenUnitWithAttachment(mockUnit1Attachment);
    unit2();
    final List<String> steps =
        getStepBuilder()
            .attackingUnits(List.of(unit1))
            .defendingUnits(List.of(unit2))
            .isBattleSiteWater(false)
            .build()
            .get();
```
In the above, it's pretty clear the usage of 'unitAttachment' is a side-effect somewhere. Following the above pattern should eventually have unitAttachment become a parameter or a return value of some setup method, so at least you could see where it is configured (which should make the test a bit more self-explanatory and self-contained).


WDYT of this approach?

Reply:
Interesting idea.  I just recently tried out a different approach where I did something like:

```
private Unit createUnit() {
   Unit unit = mock(Unit.class);
   UnitType unitType = mock(UnitType.class);
   when(unit.getType).thenReturn(unitType);
   when(unitType.getAttachment(anyString()).thenReturn(unitAttachment);
}

Reply:
The only criticism I would have for that would be this line:
```
unit.getAttachment("");
```

It takes some guess work why the empty string returns something useful. At this juncture, IMO the simplest would be to have a helper method that would allow you to write something like: 
```
Unit aaUnit = givenUnitWithAaType("AntiAirGun");
```